Support more platforms

djdv opened the pull request:
We're using this in go-ipfs (https://github.com/ipfs/go-ipfs/pull/6612), and our CI builds on a range of platforms.
I've added definitions and some methods for translating between native system structures and library standards, that may be useful here. They at least allow you to build the package on the platform, but only Windows and Linux are being tested fully at the moment.

I also attempted to port vecnet's recvmsg methods.
On Windows, this should be possible, however the implementation in this patchset seems to fail. `WSARecvMsg` is returning `WSAWSAEINVAL`. Documentation says "The socket has not been bound (with bind, for example).", which is strange because if I try to bind to the handle returned from Go, I'm told we're already bound, so this needs some investigating.

Likewise, I found it odd that Go doesn't define `syscall.SYS_RECVMSG` on linux_386, only amd64.
but it does have a method for it https://golang.org/pkg/syscall/?GOOS=linux&GOARCH=386#Recvmsg
So this needs some looking at as well.

I included everything just for visibility, but we can split anything out if desired and/or try to fix the broken bits.

hugelgupf commented:
Hey, sorry for the delay, I totally missed the email notifying me of this. Looking now...

hugelgupf commented on vecnet/vecnet_linux_arm.go:
```diff
@@ -0,0 +1,109 @@
+// Copyright 2018 The gVisor Authors.
```
how are these different on arm,arm64? i don't see the diff from x86, but it's also a bit needle-in-haystack

djdv replied:
Specifically it's the syscall iovec's length size.
>Len:  uint32(l)
Len:  uint64(l)
msg.Iovlen = uint32(len(iovecs))
msg.Iovlen = uint64(len(iovecs))

There may be a better way to reduce this though, this was some quick copy paste hackery.

hugelgupf replied:
#15 has examples of how to make this easier

djdv commented on localfs/system_windows.go:
```diff
@@ -0,0 +1,40 @@
+	}
+	fd, err := windows.CreateFile(pathPtr, access, sharemode, nil, createmode, attribute, 0)
+	if err != nil {
+		fmt.Println("hit 2: ", pathPtr, "\n", path, "\n", windows.GetLastError())
```
The prints should be removed at the very least but I was reluctant to change the function signature (appending an error to the return).
If Go was able to retrieve a stat for the file already, it's highly unlikely we'd be unable to access it again here, using the system native methods.
That being said, a panic is a panic, so maybe it would be best to change ` func localToQid(path string, info os.FileInfo) uint64` to ` func localToQid(path string, info os.FileInfo) (uint64, error)`

djdv commented on sys/stat_darwin.go:
```diff
@@ -0,0 +1,27 @@
+package sys
```
We can likely fold freebsd and darwin without problems.
But the issues with Linux and OpenBSD_amd64 is that the native blocksize is defined as an int32 on OpenBSD_amd64, where it's defined as uint32 on most other platforms.
It's a uin32 on OpenBSD_386 though.
https://golang.org/pkg/syscall/?GOOS=openbsd&GOARCH=386#Stat_t
https://golang.org/pkg/syscall/?GOOS=openbsd&GOARCH=amd64#Stat_t

Maybe we condense into stat.go, stat_bsd.go, and stat_openbsd_amd64.go.
I'll mess around with these and throw them at the build CI.

hugelgupf replied:
Wow. Lol.

hugelgupf commented:
btw, #15 added support for a bunch more architectures on linux. you need not support all the same archs in windows.

djdv commented:
@hugelgupf 
Sorry for the delay. I pushed up some changes.
- removed some of the redundancy in the `sys` pkg,
- added platforms on top of #15
- removed the Windows recvmsg implementation (since it didn't work).
- `localToQid` now returns an error

Besides the format standards the CI bot doesn't like, this should be good for another pass if you could.

I'm not sure where best to put the `sys` package, and if things inside of it should be exported or internalized though.
Also not sure if anything else should be removed from this (like the go.mod files, etc.).

djdv commented on localfs/readdir.go:
```diff
@@ -0,0 +1,50 @@
+		if err != nil {
+			if err == io.EOF {
+				err = nil // we want to return an empty RREADDIR, not an RERROR
+			}
```
Not sure if we should be consuming this error or returning it.

hugelgupf replied:
i think this should go something like this:

```go
if err == io.EOF {
  return p9Ents, nil
} else if err != nil {
  return nil, err
}
```

djdv replied:
Done: https://github.com/hugelgupf/p9/pull/12/commits/0ecadc9ec01c175c48af3d85aec6c93d07382bab

hugelgupf commented on localfs/readdir.go:
```diff
@@ -0,0 +1,50 @@
+		cursor++
+
+		// skip processing it if it was not requested
+		if offset > cursor {
```
```go
// cursor \in (offset, offset+count)
if cursor < offset || cursor > offset+count {
  continue
}
```

djdv replied:
Done: https://github.com/hugelgupf/p9/pull/12/commits/395f1a75a923b2d8eddd5abaeab1381720f2a8a2

hugelgupf commented on sys/stat_windows.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+//TODO: copied from pkg p9
+// we should probably migrate the OS methods from p9 into sys
```
hmmm. probably, yeah.

I'll probably end up making a copy of gVisor's pkg/abi/linux for this at some point.

hugelgupf replied:
sys sgtm for now, though

hugelgupf commented on vecnet/vecnet_linux_386.go:
```diff
@@ -0,0 +1,23 @@
+	"syscall"
+)
+
+//TODO / FIXME: pkg syscall doesn't seem to define constants needed to make the Recvmsg syscall directly (through .Syscall())
```
do you need this file? it should compile and just not use the optimized version if you remove it, i thiiink?

djdv replied:
Good call. Looks like There was some more that could be reduced as well.
https://github.com/hugelgupf/p9/pull/12/commits/3ac24bae1de6a71fdee2a9be1a2c0f77b7782caa

djdv commented:
New changes:
- return appropriate values in `Readdir` https://github.com/hugelgupf/p9/pull/12/commits/0ecadc9ec01c175c48af3d85aec6c93d07382bab
- better constrain `Readdir` entry processing https://github.com/hugelgupf/p9/pull/12/commits/395f1a75a923b2d8eddd5abaeab1381720f2a8a2 
- consolidate sys/stat_t transforms and fix building on DragonflyBSD https://github.com/hugelgupf/p9/pull/12/commits/3ac24bae1de6a71fdee2a9be1a2c0f77b7782caa
(`stat_linux.go`, `stat_solaris.go`, and `stat_dragonfly.go` turned into `stat_standard.go`)

Tested with `gox -output="NUL" -os="windows linux freebsd dragonfly darwin netbsd openbsd solaris" -arch="386 amd64 amd64p32 arm arm64 ppc64 ppc64le s390x mips mipsle mip64 mips64le"` against the `local_server`.
 Builds on all except `dragonfly/386` (nonexistant GOOS/GOARCH pair) and `darwin/arm, arm64` (I'm lacking darwin arm toolchains).
